### PR TITLE
fix: DATASET_READ_TABLE read permissions

### PR DESCRIPTION
**Pull request by SofiaSazonova**
### Feature or Bugfix
- Bugfix

### Detail
- backfill DATASET_READ_TABLE permissions
- delete this permissions, when dataset tables are revoked or deteled 
- 
### Relates
- #1173

### Security
Please answer the questions below briefly where applicable, or write `N/A`. Based on
[OWASP 10](https://owasp.org/Top10/en/).

- Does this PR introduce or modify any input fields or queries - this includes
fetching data from storage outside the application (e.g. a database, an S3 bucket)?
  - Is the input sanitized?
  - What precautions are you taking before deserializing the data you consume?
  - Is injection prevented by parametrizing queries?
  - Have you ensured no `eval` or similar functions are used?
- Does this PR introduce any functionality or component that requires authorization?
  - How have you ensured it respects the existing AuthN/AuthZ mechanisms?
  - Are you logging failed auth attempts?
- Are you using or adding any cryptographic features?
  - Do you use a standard proven implementations?
  - Are the used keys controlled by the customer? Where are they stored?
- Are you introducing any new policies/roles/users?
  - Have you used the least-privilege principle? How?

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by SofiaSazonova** [comment]
`DATASET_TABLE_READ = [GET_DATASET_TABLE, PREVIEW_DATASET_TABLE]`

**Checks**
- [x] check `DATASET_TABLE_READ` against the table when we execute the `get_table` service function
`DatasetTableService, line 41`
- [x] check `PREVIEW_DATASET_TABLE` against the table when we execute the `preview` service function
`DatasetTableService line 98`

**Give permissions**
- [x] attach `DATASET_TABLE_READ` permission to the dataset owner and stewards when the table is created
`DatasetTableService line 146`
- [x] attach `DATASET_TABLE_READ` permission to the new stewards when the table is transferred to new stewards
`DatasetService line 520`
- [x] attach `DATASET_TABLE_READ` permissions to the share requester group when the table is shared
`ShareObjectService line 317`

**Remove permissions**
- [x] remove all existing resource policies for the table permissions when the table is deleted
`DatasetTableService line 80`
- [x] remove `DATASET_TABLE_READ` permissions to old stewards when the table is transferred to new stewards
`DatasetService line 512`
- [x] remove `DATASET_TABLE_READ` permissions to the share requester group when the table is revoked
`ShareItemService line 107`

**Database chages**
- [x] Migration script that creates the new permission type and creates records backfilling the existing datasets and shares corresponding permissions.

**Comment by SofiaSazonova** [comment]
Also solves #1171 steps 2 and 3

**Comment by SofiaSazonova** [comment]
After conversation with @noah-paige I understood that for step 1 I have done the opposite thing. So, reverting it

File: backend/dataall/core/permissions/db/resource_policy/resource_policy_repositories.py [hunk 1]
```diff
@@ -71,16 +70,35 @@ def has_group_resource_permission(
...
-            .first()
+    def query_all_resource_policies(session, group_uri: str, resource_uri: str, resource_type: str = None):
+        resource_policy = session.query(ResourcePolicy).filter(
+            ResourcePolicy.principalId == group_uri,
+            ResourcePolicy.resourceUri == resource_uri,
         )
+        if group_uri is not None:
+            resource_policy = resource_policy.filter(
+                ResourcePolicy.principalId == group_uri,
+            )
```
Comment (noah-paige) [inline backend/dataall/core/permissions/db/resource_policy/resource_policy_repositories.py:81]: We are filtering on `ResourcePolicy.principalId == group_uri` on L75 and L79-81

Can we remove L75 and keep logic in L78-81?

File: backend/dataall/modules/datasets/services/dataset_table_service.py [hunk 2]
```diff
@@ -76,6 +77,7 @@ def delete_table(uri: str):
 
             ShareObjectRepository.delete_shares(session, table.tableUri)
             DatasetTableRepository.delete(session, table)
+            DatasetTableService._delete_dataset_table_read_permission(session, uri)
```
Comment (noah-paige) [inline backend/dataall/modules/datasets/services/dataset_table_service.py:80]: This works for when a user manually goes to delete a table but something to think about is how best should we handle table syncs (I am not sure the correct answer here). For instance:

- Either when a user starts a manual sync or the scheduled table_syncer ECS task is run (ultimately both run `DatasetTableService.sync_existing_tables()` at some point)
- If the glue table exists in data.all but does not exist in the glue response --> we update table status to `Deleted`
- On the UI we no longer show those tables by filtering for `!= Deleted` (ref: `DatasetService.paginated_dataset_tables()`
- But these tables still do have associated permission records for who should be able to access the table
  - And if we do clean them up right away but the table DOES still exist on Glue (i.e. some other error in API Response or similar returns 0 tables incorrectly) then we have potentially just broken existing access or shares if the next time a user hits `Sync` the table re-appears

Some ideas of what we can do:
1. Do nothing. There will be stale tables and permission records in RDS but it avoids risk of removing permissions inappropriately and should not greatly affect logic of data.all
2. Implement some type of garbage collection. Delete tables and associated shares on those tables after the tables have been in status of `Deleted` for some extended period of time (i.e. 30 days)
3. As soon as a table status gets updated to `Deleted` still show it on the UI but with a `Deprecated` Flag and with the only option for the user to do is to delete the table + clean up shares (pre-req to table delete already)
    - Remove from Catalog (already done by sync), prevent all new shares, only can revoke share and clean-up / delete of table
    - If next sync restores the glue table back to `InSync` allow for normal activity again (no longer Deprecated)


TLDR - After writing all the above may be best to open this up as a new issue to be tackled later as well but deserves some thought!

File: backend/migrations/versions/458572580709_remove_dataset_table_read_permissions_from_env_admins.py [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+                        and_(
+                            ShareObject.principalType == PrincipalType.Group.value,
+                            ShareObject.principalId == env_admin_group,
+                            ShareObjectItem.shareItemUri == table.tableUri,
+                            ShareObjectItem.itemType == ShareableType.Table.value,
+                            ShareObjectItem.status.in_(
+                                [
+                                    ShareItemStatus.Share_Approved.value,
+                                    ShareItemStatus.Share_Succeeded.value,
+                                    ShareItemStatus.Share_In_Progress.value,
```
Comment (noah-paige) [inline backend/migrations/versions/458572580709_remove_dataset_table_read_permissions_from_env_admins.py:71]: ore Revoke_Failed status as well right?

File: backend/migrations/versions/458572580709_remove_dataset_table_read_permissions_from_env_admins.py [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+def upgrade():
+    session = get_session()
+
+    datasets: [Dataset] = session.query(Dataset).all()
+    for dataset in datasets:
+        environment = EnvironmentService.get_environment_by_uri(session, uri=dataset.environmentUri)
+        env_admin_group = environment.SamlGroupName
+
+        # if envAdmis is also Dataset admin, no need to delete permissions
+        if env_admin_group == dataset.SamlAdminGroupName:
```
Comment (noah-paige) [inline backend/migrations/versions/458572580709_remove_dataset_table_read_permissions_from_env_admins.py:48]: or dataset.steward

File: backend/migrations/versions/59ab197fe648_add_backfill_read_table_permissions.py [hunk 5]
```diff
@@ -0,0 +1,122 @@
+"""add_backfill_read_table_permissions
```
Comment (noah-paige) [inline backend/migrations/versions/59ab197fe648_add_backfill_read_table_permissions.py:1]: do we need to backfill these permissions?

File: backend/dataall/core/permissions/db/resource_policy/resource_policy_repositories.py (hunk 1 above)
Comment (SofiaSazonova) [reply backend/dataall/core/permissions/db/resource_policy/resource_policy_repositories.py:81]: It's typo. Surely it's not necessary in line 75

File: backend/dataall/core/permissions/services/resource_policy_service.py [hunk 6]
```diff
@@ -75,21 +76,49 @@ def check_user_resource_permission(session, username: str, groups: [str], resour
...
-            session.delete(policy)
+        """
+        Deletes all resources policy for given group, resource_uri, resource_type
+        :param session:
+        :param group:
+        :param resource_uri:
+        :param resource_type:
+        :return:
+        """
+        policies = ResourcePolicyService.find_resource_policies(session, group, resource_uri, resource_type)
```
[the remark this answers is not in the document]
Comment (SofiaSazonova) [reply backend/dataall/core/permissions/services/resource_policy_service.py:112]: Yes, it's incapsulated in find_resource_policies.
So, if we can find, we can delete.

File: backend/migrations/versions/458572580709_remove_dataset_table_read_permissions_from_env_admins.py (hunk 4 above)
Comment (SofiaSazonova) [reply backend/migrations/versions/458572580709_remove_dataset_table_read_permissions_from_env_admins.py:48]: Sure

File: backend/dataall/modules/datasets/services/dataset_table_service.py (hunk 2 above)
Comment (SofiaSazonova) [reply backend/dataall/modules/datasets/services/dataset_table_service.py:80]: Let's have a new issue for this!

File: backend/dataall/core/permissions/services/resource_policy_service.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (noah-paige) [reply backend/dataall/core/permissions/services/resource_policy_service.py:112]: silly of me - i see now

File: backend/dataall/modules/datasets/services/dataset_table_service.py (hunk 2 above)
Comment (noah-paige) [reply backend/dataall/modules/datasets/services/dataset_table_service.py:80]: New Issue Here - https://github.com/data-dot-all/dataall/issues/1246

File: backend/migrations/versions/59ab197fe648_add_backfill_read_table_permissions.py (hunk 5 above)
Comment (noah-paige) [reply backend/migrations/versions/59ab197fe648_add_backfill_read_table_permissions.py:1]: Don't we already assign `DATASET_TABLE_READ` permission on creates of tables and on share approvals? And `DATASET_TABLE_READ` already included `GET_DATASET_TABLE` and `PREVIEW_DATASET_TABLE`?

So were we not already covered with assigning these permissions - we just were not originally checking them? Maybe I am missing something?

Comment (SofiaSazonova) [reply backend/migrations/versions/59ab197fe648_add_backfill_read_table_permissions.py:1]: Yes, I checked in DB. This permissions are already in place

**Comment by SofiaSazonova** [comment]
@noah-paige good catch!
I will try to solve in the next commit

**Comment by SofiaSazonova** [comment]
@dlpzx @noah-paige I placed attachement/deletion of permissions inside share processor workers. So, the permissions are only granted|revoked when the share item is processed successfully.

Upon success of the share the permissions are granted to all items at once, so, I better make it item per item. But have a look at this version, please

File: backend/dataall/modules/dataset_sharing/services/data_sharing_service.py [hunk 7]
```diff
@@ -176,6 +177,15 @@ def approve_share(cls, engine: Engine, share_uri: str) -> bool:
                 new_share_state = share_sm.run_transition(ShareObjectActions.Finish.value)
                 share_sm.update_state(session, share, new_share_state)
 
+                if share_successful:
```
Comment (noah-paige) [inline backend/dataall/modules/dataset_sharing/services/data_sharing_service.py:180]: I think we still run into issues here in the following scenario:

- I approve share for 1 table and 1 folder
- 1 table succeeds but 1 folder fails (no permissions attached to requestor)
- I delete a re-submit folder and it succeeds (requestor gets permissions only to folder NOT to table)

May need to attach permissions 1 by 1 to each item to avoid the above

File: backend/dataall/modules/dataset_sharing/services/share_processors/s3_access_point_process_share.py [hunk 8]
```diff
@@ -185,7 +185,6 @@ def process_revoked_shares(
                     if not dataset.imported or dataset.importedKmsKey:
                         removing_folder.delete_dataset_bucket_key_policy(dataset=dataset)
                 new_state = revoked_item_SM.run_transition(ShareItemActions.Success.value)
-                revoked_item_SM.update_state_single_item(session, removing_item, new_state)
```
Comment (noah-paige) [inline backend/dataall/modules/dataset_sharing/services/share_processors/s3_access_point_process_share.py:188]: Why are we removing this line?

File: tests/modules/datasets/tasks/test_dataset_tables_sync.py [hunk 9]
```diff
@@ -1,4 +1,5 @@
 from unittest.mock import MagicMock
+from dataall.base.context import set_context, RequestContext
```
Comment (noah-paige) [inline tests/modules/datasets/tasks/test_dataset_tables_sync.py:2]: do we need these imports?

File: backend/migrations/versions/458572580709_remove_dataset_table_read_permissions_from_env_admins.py [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+from dataall.core.environment.services.environment_service import EnvironmentService
+from dataall.core.permissions.services.resource_policy_service import ResourcePolicyService
+
+from dataall.modules.datasets_base.db.dataset_models import DatasetTable, Dataset
+from dataall.modules.dataset_sharing.db.share_object_models import ShareObject, ShareObjectItem
+from dataall.modules.dataset_sharing.services.dataset_sharing_enums import ShareItemStatus, ShareableType, PrincipalType
+
+# revision identifiers, used by Alembic.
+revision = '458572580709'
+down_revision = 'c6d01930179d'
```
Comment (noah-paige) [inline backend/migrations/versions/458572580709_remove_dataset_table_read_permissions_from_env_admins.py:22]: lets update with `a991ac7a85a2` as `down_revision` from migration script in PR https://github.com/data-dot-all/dataall/pull/1237

Comment (SofiaSazonova) [reply backend/migrations/versions/458572580709_remove_dataset_table_read_permissions_from_env_admins.py]: migrations merged

File: backend/dataall/modules/dataset_sharing/services/share_processors/s3_access_point_process_share.py (hunk 8 above)
Comment (SofiaSazonova) [reply backend/dataall/modules/dataset_sharing/services/share_processors/s3_access_point_process_share.py]: brought back

File: backend/dataall/modules/dataset_sharing/services/data_sharing_service.py (hunk 7 above)
Comment (SofiaSazonova) [reply backend/dataall/modules/dataset_sharing/services/data_sharing_service.py]: done

File: tests/modules/datasets/tasks/test_dataset_tables_sync.py (hunk 9 above)
Comment (noah-paige) [reply tests/modules/datasets/tasks/test_dataset_tables_sync.py:2]: did you push commit to remove this? Still see in after pulling latest changes

**Review by noah-paige — APPROVED** [review]
Tested 
- migration scripts
- adding and removing permission for table shares
- deleting tables + removing associated permissions (+ re-sync)

All looks good, approving
